src/distilabel/utils/serialization_v2.py

```diff
@@ -0,0 +1,231 @@
+
+    Args:
+        template (Dict[str, Any]): Dict containing the template, the dict serialized.
+
+    Returns:
+        Generic[T]: Instance contained in the template
+    """
+    type_info = class_.pop("_type_info_")
+    if "_type_info_" in type_info:
+        # There is a nested type_info, load the class recursively
+        type_info = load_from_dict(type_info)
+
+    cls = _get_class(type_info["module"], type_info["name"])
+    instance = cls(**class_)
+    return instance
+
+
+def write_json(filename: Path, data: Dict[str, Any]) -> None:
+    """Writes a json file to the given path, creates the parent dir.
+
+    Args:
+        filename (Path): Name of the file.
+        data (Dict[str, Any]): Dict to be written as json.
+    """
+    filename.parent.mkdir(parents=True, exist_ok=True)
+    with open(filename, "w") as file:
+        json.dump(data, file, indent=2)
+
+
+def read_json(filename: Path) -> Dict[str, Any]:
+    """Read a json file from disk.
+
+    Args:
+        filename (Path): Name of the json file.
+
+    Returns:
+        Dict[str, Any]: Dict containing the json data.
+    """
+    with open(filename, "r") as file:
+        return json.load(file)
+
+
+def write_yaml(filename: Path, data: Dict[str, Any]) -> None:
+    """Writes a yaml file to the given path, creates the parent dir.
+
+    Args:
+        filename (Path): Name of the file.
+        data (Dict[str, Any]): Dict to be written as yaml.
+    """
+    filename.parent.mkdir(parents=True, exist_ok=True)
+    with open(filename, "w") as file:
+        yaml.dump(data, file, default_flow_style=False)
+
+
+def read_yaml(filename: Path) -> Dict[str, Any]:
+    """Read a yaml file from disk.
+
+    Args:
+        filename (Path): Name of the yaml file.
+
+    Returns:
+        Dict[str, Any]: Dict containing the json data.
+    """
+    with open(filename, "r") as file:
+        return yaml.load(file, Loader=yaml.FullLoader)
+
+
+class _Serializable:
+    """Base class for serializable classes. It provides the means to serialize and deserialize."""
+
+    _type_info_: Dict[str, Any] = {}
+
+    def _model_dump(self, obj: Any, **kwargs: Any) -> Dict[str, Any]:
+        """Method in charge of serializing the object.
+
+        This method works for pydantic models (the classes that inherit from pydantic's
+        BaseModel).
+
+        The signature must be respected, `obj` represents the object to serialize and `kwargs` are
+        the optional parameters for the method used.
+        For example in the case of a pydantic model, the method `model_dump` is used,
+        but for the DAG class, we networkx to obtain the information.
+
+        Args:
+            obj (Any): The object to be serialized.
+
+        Returns:
+            Dict[str, Any]: Data to be dumped for the object.
+        """
+        # NOTE(plaguss): This should be reimplemented for some cases, so maybe we have to enforce it via ABC?.
+        # It will work out of the box for pydantic's BaseModels.
```

Comment:
FYI all the steps will be `pydantic.BaseModel` as those will need to inherit from `Step`, so we're safe assuming everything's a `pydantic.BaseModel` AFAIK cc @gabrielmbmb

Reply:
The comment was related to the classes we will serialize other than the `Step`, the `DAG` for example has to update this method. But for the general case of the `Step` that works, so I guess I wanted that method to let as some freedom to tweak the process without writing to much extra code.